File: leo2-ui/src/main/java/org/deku/leo2/fx/ModuleController.java
package org.deku.leo2.fx;

import javafx.fxml.FXMLLoader;
import javafx.scene.Node;
import org.deku.leo2.Main;

/**
 * User interface module
 *
 * Created by masc on 23.09.14.
 */
public abstract class ModuleController extends Controller {
    protected Node mNode;

    public static <T extends ModuleController> T fromFxml(String path) {
        FXMLLoader fxml = Main.instance().loadFxPane(path);
        T m = (T)fxml.getController();
        m.mNode = fxml.getRoot();
        return m;
    }

    public Node getNode() {
        return mNode;
    }

    public abstract String getTitle();
}

File: leoz-ui/src/main/java/org/deku/leoz/ui/fx/ModuleController.kt
package org.deku.leoz.ui.fx

import javafx.scene.image.Image
import org.deku.leoz.ui.event.BusyNotifier
import org.deku.leoz.ui.event.ErrorNotifier
import org.deku.leoz.ui.event.Event
import io.reactivex.subjects.PublishSubject

/**
 * Base class for leoz user interface modules

 * Created by masc on 23.09.14.
 */
abstract class ModuleController : Controller(), BusyNotifier, ErrorNotifier {
    /**
     * Module title
     * @return
     */
    abstract val title: String

    /**
     * Module title image
     * @return
     */
    abstract val titleImage: Image

    // BusyNotifier implementation
    override val ovBusy by lazy { PublishSubject.create<Event<Boolean>>() }

    // ErrorNotifier implementation
    override val ovError by lazy { PublishSubject.create<Exception>() }
}
